client/post-editor/editor-sharing/publicize-services.jsx

```diff
@@ -1,59 +1,72 @@
+import React, { Component, PropTypes } from 'react';
+import { connect } from 'react-redux';
+import { filter, uniqBy } from 'lodash';
 
 /**
  * Internal dependencies
  */
-var serviceConnections = require( 'my-sites/sharing/connections/service-connections' ),
-	EditorSharingPublicizeConnection = require( './publicize-connection' );
+import EditorSharingPublicizeConnection from './publicize-connection';
+import { getCurrentUserId } from 'state/current-user/selectors';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getSiteUserConnections } from 'state/sharing/publicize/selectors';
 
-module.exports = React.createClass( {
-	displayName: 'EditorSharingPublicizeServices',
+class EditorSharingPublicizeServices extends Component {
+	static propTypes = {
+		post: PropTypes.object,
+		connections: PropTypes.array.isRequired,
+		newConnectionPopup: PropTypes.func.isRequired
+	};
 
-	propTypes: {
-		post: React.PropTypes.object,
-		siteId: React.PropTypes.number.isRequired,
-		connections: React.PropTypes.array.isRequired,
-		newConnectionPopup: React.PropTypes.func.isRequired
-	},
+	static defaultProps = {
+		post: Object.freeze( {} ),
+	};
 
-	renderServices: function() {
-		var services = serviceConnections.getServicesFromConnections( this.props.connections );
-
-		return services.map( function( service ) {
-			return (
-				<li key={ service.ID } className="editor-sharing__publicize-service">
-					<h5 className="editor-sharing__publicize-service-heading">{ service.label }</h5>
-					{ this.renderConnections( service.ID ) }
-				</li>
-			);
-		}, this );
-	},
+	renderServices() {
+		const services = uniqBy( this.props.connections.map( ( { label, service } ) => ( { label, service } ) ), 'service' );
 
-	renderConnections: function( serviceName ) {
-		const connections = serviceConnections.getConnectionsAvailableToCurrentUser(
-			serviceName,
-			this.props.connections
+		return services.map( ( { label, service } ) =>
+			<li key={ service } className="editor-sharing__publicize-service">
+				<h5 className="editor-sharing__publicize-service-heading">{ label }</h5>
+				{ this.renderConnections( service ) }
+			</li>
 		);
+	}
 
-		return connections.map( function( connection ) {
-			return (
-				<EditorSharingPublicizeConnection
-					key={ connection.ID }
-					post={ this.props.post }
-					connection={ connection }
-					onRefresh={ this.props.newConnectionPopup } />
-			);
-		}, this );
-	},
+	/**
+	 * Displays Publicize connections for the passed service.
+	 *
+	 * @param {String} service Slug of Keyring service.
+	 * @return {Object|DOMElement} React Element.
+	 */
+	renderConnections( service ) {
+		return filter( this.props.connections, { service } ).map( ( connection ) =>
+			<EditorSharingPublicizeConnection
+				key={ connection.ID }
+				post={ this.props.post }
+				connection={ connection }
+				onRefresh={ this.props.newConnectionPopup } />
+		);
+	}
 
-	render: function() {
+	render() {
 		return (
```

Comment:
>Here I wondered how other reviewers would feel about loops two levels deep :)

To be fair, it should be framed in the context of vs. multiple render functions. There's a third alternative which would be intermediate components that handle specific behaviors (say, one for the list item with the label.

>Aside: What makes you prefer `_.map` over `[].map()`?

In the general case, I might use it because Lodash tends to perform much better than the native equivalents because it "takes shortcuts" vs. complete spec compliance.

In this case, it was more a point of consistency. The first `map` is actually mapping an object, which would otherwise not be possible with native iterators (at least, not as elegantly).

>What's the benefit of keying connections with their label instead of accessing that property directly within the loop?

Because at the point we use the label, we only have access to the label, an array of connections for that label, and all available connections. If you were to key by service instead, I suppose you could pick the label from the first entry in the `connections` value, e.g. `connections[ service ][ 0 ].label`. But again, the idea is that by losing the distinction between label and service, we happily accept that all connections with the same label should be grouped together, even in the unlikely scenario that their service property be different.